### PR TITLE
ISPN-16700 Combine the infinispan-hotrod module with the infinispan-client-hotrod-new

### PR DESCRIPTION
**Pull request by jabolina**
https://issues.redhat.com/browse/ISPN-16700

I've merged the whole API. The API is there but some methods need the actual implementation. The sync/async caches should be fine but are missing query, listener, and streaming. To get it working I had to change a few operations to return a `CacheEntry<K, V>` in a few places. This means we must hold the key to pass to the cache entry.

The counters, locks, and multimap need an actual implementation. We can get this one in and later implement the methods. This approach would help during the review as well.

### CODE REVIEW

**Comment by jabolina** [comment]
The big thing is the missing tests! The tests on the infinispan-hotrod module were using JUnit, so I didn't copy them. This means we don't have actual tests for the new API.

**Comment by wburns** [comment]
Maybe move the tests into a tests only module until we can move the testNG to it and then merge it back in?

**Comment by jabolina** [comment]
Yeah, I think that would be best. I'll keep the infinispan-hotrod module for testing only.

File: client/hotrod-client-new/src/main/java/org/infinispan/client/hotrod/impl/operations/AbstractKeyHolderOperation.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+import org.infinispan.api.common.CacheEntry;
+import org.infinispan.client.hotrod.impl.InternalRemoteCache;
+import org.infinispan.client.hotrod.impl.VersionedOperationResponse;
+import org.infinispan.client.hotrod.impl.protocol.Codec;
+import org.infinispan.client.hotrod.impl.protocol.HotRodConstants;
+
+import io.netty.buffer.ByteBuf;
+
+public abstract class AbstractKeyHolderOperation<K, V> extends AbstractKeyOperation<V> {
```
Comment (jabolina) [inline client/hotrod-client-new/src/main/java/org/infinispan/client/hotrod/impl/operations/AbstractKeyHolderOperation.java:11]: I've added this class to the hierarchy to hold the key for the specific operations in the new API. I didn't want to change `AbstractKeyOperation` as that seems to affect many more classes.

Let me know if you can think of another approach for this.

Comment (wburns) [reply client/hotrod-client-new/src/main/java/org/infinispan/client/hotrod/impl/operations/AbstractKeyHolderOperation.java:11]: I don't think this is required at all. I believe I sent it to you via message, but I highly recommend just changing the affected commands to just return a `MetadataValue` just like `GetWithMetadataOperation` does today (ignoring the retry part). Then the RemoteCacheImpl can convert the `MetadataValue` to a `CacheEntry` in a `thenApply` as it already has the key from the request. Then we don't need to add this class.

File: client/hotrod-client-new/src/main/java/org/infinispan/client/hotrod/impl/RemoteCacheImpl.java [hunk 2]
```diff
@@ -821,4 +856,227 @@ public CompletionStage<Void> updateBloomFilter() {
...
+
+   @Override
+   public CompletionStage<CacheConfiguration> configuration() {
+      return CompletableFutures.completedNull();
+   }
+
+   @Override
+   public CompletionStage<V> get(K key, CacheOptions options) {
+      assertRemoteCacheManagerIsStarted();
+      HotRodOperation<V> op = new AdvancedHotRodOperation<>(operationsFactory.newGetOperation(key), options);
```
Comment (wburns) [inline client/hotrod-client-new/src/main/java/org/infinispan/client/hotrod/impl/RemoteCacheImpl.java:868]: I like the use of AdvancedHotRodOperation here :) Just shows how nice the flexibility is.

File: client/hotrod-client-new/src/main/java/org/infinispan/client/hotrod/impl/RemoteCacheImpl.java [hunk 3]
```diff
@@ -821,4 +856,227 @@ public CompletionStage<Void> updateBloomFilter() {
...
+      long lifespan = CacheOptionsUtil.lifespan(options, defaultLifespan, TimeUnit.MILLISECONDS);
+      long maxIdle = CacheOptionsUtil.maxIdle(options, defaultMaxIdleTime, TimeUnit.MILLISECONDS);
+      HotRodOperation<CacheEntry<K, V>> op = operationsFactory.newPutKeyValueOperation(key, value, lifespan, TimeUnit.MILLISECONDS,
+            maxIdle, TimeUnit.MILLISECONDS);
+      return dispatcher.execute(new AdvancedHotRodOperation<>(op, options, PrivateHotRodFlag.FORCE_RETURN_VALUE.getFlagInt()));
+   }
+
+   @Override
+   public CompletionStage<Void> set(K key, V value, CacheWriteOptions options) {
+      return put(key, value, options).thenApply(CompletableFutures.toNullFunction());
```
Comment (wburns) [inline client/hotrod-client-new/src/main/java/org/infinispan/client/hotrod/impl/RemoteCacheImpl.java:905]: We don't want FORCE_RETURN_VALUE in this case.

**Comment by wburns** [comment]
I won't try to make a big review comment as Github just can't seem to keep up with the diff so I will just be adding one off comments if I see anymore.

File: client/hotrod-client-new/src/main/java/org/infinispan/client/hotrod/HotRodAsyncCache.java [hunk 4]
```diff
@@ -166,7 +166,7 @@ public CompletionStage<Void> clear(CacheOptions options) {
 
    @Override
    public <R> AsyncQuery<K, V, R> query(String query, CacheOptions options) {
-      return new HotRodAsyncQuery(query, options);
+      return new HotRodAsyncQuery<>();
```
Comment (wburns) [inline client/hotrod-client-new/src/main/java/org/infinispan/client/hotrod/HotRodAsyncCache.java:169]: I would say just throw UnsupportedOperationException as well until we actually implement this.

File: client/hotrod-client-new/src/main/java/org/infinispan/client/hotrod/HotRodAsyncCaches.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+   @Override
+   public <K, V> CompletionStage<AsyncCache<K, V>> create(String name, String template) {
+      hotrod.cacheManager.getConfiguration().addRemoteCache(name, builder -> builder.templateName(template));
+      return get(name);
+   }
+
+   @Override
+   public CompletionStage<Void> remove(String name) {
+      hotrod.cacheManager.administration().removeCache(name);
```
Comment (wburns) [inline client/hotrod-client-new/src/main/java/org/infinispan/client/hotrod/HotRodAsyncCaches.java:44]: Sadly all of these methods are really non blocking.. but they are just doing an await call.

Comment (jabolina) [reply client/hotrod-client-new/src/main/java/org/infinispan/client/hotrod/HotRodAsyncCaches.java:44]: Yeah. I had another branch change the RemoteCacheManagerImpl to return a CompletableFuture, but I didn't pull it into this PR so the change wouldn't be lost in the huge diff.

File: client/hotrod-client-new/src/main/java/org/infinispan/client/hotrod/impl/AdvancedRemoteCache.java [hunk 6]
```diff
@@ -19,35 +16,15 @@
...
-import org.reactivestreams.Publisher;
-
-/**
- * @since 14.0
- **/
-public interface RemoteCache<K, V> extends AutoCloseable {
-   CompletionStage<CacheConfiguration> configuration();
 
-   HotRodTransport getHotRodTransport();
+public interface AdvancedRemoteCache<K, V> {
```
[the remark this answers is not in the document]
Comment (jabolina) [reply client/hotrod-client-new/src/main/java/org/infinispan/client/hotrod/impl/AdvancedRemoteCache.java:20]: Sounds good. I was debating that.

File: client/hotrod-client-new/src/main/java/org/infinispan/client/hotrod/HotRodAsyncCache.java (hunk 4 above)
Comment (jabolina) [reply client/hotrod-client-new/src/main/java/org/infinispan/client/hotrod/HotRodAsyncCache.java:171]: I kept the classes from the hotrod module but added UnsupportedOperationException to all the unimplemented methods.

**Comment by jabolina** [comment]
Updated with all the suggestions.

File: client/hotrod-client-new/src/main/java/org/infinispan/client/hotrod/impl/RemoteCacheImpl.java (hunk 2 above)
Comment (jabolina) [reply client/hotrod-client-new/src/main/java/org/infinispan/client/hotrod/impl/RemoteCacheImpl.java:860]: This approach with the delegate helped SO MUCH in this PR.

File: client/hotrod-client-new/src/main/java/org/infinispan/client/hotrod/impl/operations/RoutingCacheOperationsFactory.java [hunk 7]
```diff
@@ -19,42 +20,42 @@ public <V> HotRodOperation<V> newGetOperation(Object key) {
...
 
    @Override
-   public <V, K> HotRodOperation<V> newPutIfAbsentOperation(K key, V value, long lifespan, TimeUnit lifespanUnit, long maxIdleTime, TimeUnit maxIdleTimeUnit, Flag... flags) {
+   public <K, V> HotRodOperation<MetadataValue<V>> newPutIfAbsentOperation(K key, V value, long lifespan, TimeUnit lifespanUnit, long maxIdleTime, TimeUnit maxIdleTimeUnit, Flag... flags) {
       return new RoutingObjectOperation<>(super.newPutIfAbsentOperation(key, value, lifespan, lifespanUnit, maxIdleTime, maxIdleTimeUnit, flags), key);
    }
 
    @Override
-   public <V, K> HotRodOperation<V> newReplaceOperation(K key, V valueBytes, long lifespan, TimeUnit lifespanUnit, long maxIdleTime, TimeUnit maxIdleTimeUnit) {
+   public <K, V> HotRodOperation<V> newReplaceOperation(K key, V valueBytes, long lifespan, TimeUnit lifespanUnit, long maxIdleTime, TimeUnit maxIdleTimeUnit) {
```
Comment (wburns) [inline client/hotrod-client-new/src/main/java/org/infinispan/client/hotrod/impl/operations/RoutingCacheOperationsFactory.java:48]: Shouldn't this command return a MetadataValue<V>  as well?

**Review by wburns — APPROVED** [review]
LGTM

**Comment by wburns** [comment]
Green build is crazy, just the one last question about remove. Maybe it doesn't need the MetadataValue in which case I can merge.

File: client/hotrod-client-new/src/main/java/org/infinispan/client/hotrod/impl/operations/RoutingCacheOperationsFactory.java (hunk 7 above)
Comment (jabolina) [reply client/hotrod-client-new/src/main/java/org/infinispan/client/hotrod/impl/operations/RoutingCacheOperationsFactory.java:48]: Not needed. The remove operation in the new API uses the versioned replace (ReplaceIfUnmodifiedOperation) instead.

**Comment by wburns** [comment]
Integrated into main, thanks @jabolina !

**Pull request merged**